File: swap-nodes-in-pairs/swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
void pairSwaps(ListNode* head) {
    if (head != NULL && head->next != NULL) {
        int temp = head->val;
        head->val = head->next->val;
        head->next->val = temp;
        pairSwaps(head->next->next);
    }
}
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode * x = head;
        if( head == nullptr) return head;
        if(head->next == nullptr) return head;
        // head = head->next;
        pairSwaps(x);
        return head;
    }
};

File: swap-nodes-in-pairs/swap-nodes-in-pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {    
    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next == null) return head;
        ListNode d = new ListNode(-1, head),
            pre = d,
            p = head,
            c = p.next;
        
        while(c != null) {
            ListNode forw = c.next;
            p.next = forw;
            c.next = p;
            pre.next = c;
            
            pre = p;
            p = forw;
            if(forw != null)c = forw.next; else c = null;
        }
        
        return d.next;
    }
}